rad/src/lib.rs

```diff
@@ -26,56 +29,107 @@ pub mod types;
 
 pub type Result<T> = std::result::Result<T, RadError>;
 
-pub async fn try_data_request(request: &RADRequest) -> RadonReport<RadonTypes> {
+/// The return type of any method executing the entire life cycle of a data request.
+#[derive(Debug, Serialize)]
+pub struct RADRequestExecutionReport {
+    /// Vector of reports about retrieval of data sources.
+    pub retrieve: Vec<RadonReport<RadonTypes>>,
+    /// Report about aggregation of data sources.
+    pub aggregate: RadonReport<RadonTypes>,
+    /// Report about aggregation of reports (reveals, actually).
+    pub tally: RadonReport<RadonTypes>,
+}
+
+/// Executes a data request locally.
+/// The `inputs_injection` allows for disabling the actual retrieval of the data sources and
+/// the provided strings will be fed to the retrieval scripts instead. It is therefore expected that
+/// the length of `sources_injection` matches that of `request.retrieve`.
+pub fn try_data_request(
+    request: &RADRequest,
+    settings: RadonScriptExecutionSettings,
+    inputs_injection: Option<&[&str]>,
+) -> RADRequestExecutionReport {
     let context = &mut ReportContext::default();
 
-    let retrieve_responses_fut = request
-        .retrieve
-        .iter()
-        .map(|retrieve| run_retrieval_report(retrieve));
+    let retrieve_responses = if let Some(inputs) = inputs_injection {
+        assert_eq!(inputs.len(), request.retrieve.len(), "Tried to locally run a data request with a number of injected sources different than the number of retrieval paths ({} != {})", inputs.len(), request.retrieve.len());
 
-    let retrieve_responses: Vec<RadonTypes> = futures::future::join_all(retrieve_responses_fut)
-        .await
+        let mut retrieve_responses = Vec::new();
```

Comment:
Now you don't need to manually push to the vector anymore:

```suggestion
        request.retrieve.iter().zip(inputs.iter()).map(|(retrieve, input)| {
            run_retrieval_with_data_report(
                retrieve, input, context, settings,
            )
        }).collect()
```